exec/src/klio_exec/runners/pubsub_message_manager.py

```diff
@@ -0,0 +1,284 @@
+# Copyright 2021 Spotify AB
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import logging
+import threading
+import time
+
+import apache_beam as beam
+
+from apache_beam.io.gcp import pubsub as beam_pubsub
+from apache_beam.runners.direct import direct_runner
+from apache_beam.runners.direct import transform_evaluator
+from apache_beam.utils import timestamp as beam_timestamp
+from google.cloud import pubsub as g_pubsub
+
+from klio_core.proto import klio_pb2
+
+
+ENTITY_ID_TO_ACK_ID = {}
+
+
+class PubSubKlioMessage:
+    def __init__(self, ack_id, kmsg_id):
+        self.ack_id = ack_id
+        self.kmsg_id = kmsg_id
+        self.last_extended = None
+        self.ext_duration = None
+        self.event = threading.Event()
+
+    def extend(self, duration):
+        self.last_extended = time.monotonic()
+        self.ext_duration = duration
+
+    def __repr__(self):
+        return f"PubSubKlioMessage(kmsg_id={self.kmsg_id})"
+
+
+class MessageManager:
+    DEFAULT_DEADLINE_EXTENSION = 30
+
+    def __init__(self, sub_name, heartbeat_sleep=10, manager_sleep=5):
+        self._client = g_pubsub.SubscriberClient()
+        self._sub_name = sub_name
+        self.heartbeat_sleep = heartbeat_sleep
+        self.manager_sleep = manager_sleep
+        self.messages = []
+        self.messages_lock = threading.Lock()
```

Comment:
Non-blocking just solidifying my understanding of whats going on:
So we have to make sure to have this messages lock because we have our main thread marking pmsgs in self.messages and pmsg.set() when its done processing, we have a heartbeat that accesses self.messages to give a heartbeat of what messages are still being processed as well as a MessageManager that actually does the `acking` of completed messages and thats the reason we have the lock?

Reply:
not sure if I quite understand your understanding 😆 but maybe this will help: we use a lock in this case to prevent modifying the `self.messages` list (since other parts of the code could be adding to or removing from the list at any point) in the middle of us iterating over it in the `manage` method. If we didn't, we essentially get into a circumstance where the state of the list & iteration are unknown (was a message added to this list right after I finished the last loop? was a message deleted that I'm in the middle of handling?).  So, with this, we won't iterate over the messages if we are either adding to the list or removing from the list. And we won't add or remove from the list if we're currently iterating over it. Does that make sense?

Reply:
Perf, that makes sense. Thank you!